### PR TITLE
LDAP: fix null where IUser is expected in update groups background job

### PR DESCRIPTION
**Pull request by blizzz**
This can happen when a user was already removed, but the DB table was having old information. Previously – with the old events – it did not cause a problem, because an old user id was passed. Now, an IUser object was required. Comes with tests and thus some refactoring against the  necessary methods of the job. The actual fix is the second commit.

### CODE REVIEW

**Review by skjnldsv — APPROVED** [review]
:elephant:

**Review by rullzer — APPROVED** [review]
even test.
:elephant:

File: apps/user_ldap/lib/Jobs/UpdateGroups.php [hunk 1]
```diff
@@ -79,28 +114,35 @@ public function updateGroups() {
...
+	private function getKnownGroups() {
+		if (is_array($this->groupsFromDB)) {
+			$this->groupsFromDB;
+		}
+		$qb = $this->dbc->getQueryBuilder();
+		$qb->select(['owncloudname', 'owncloudusers'])
+			->from('ldap_group_members');
+		$result = $qb->execute()->fetchAll();
+
+		$this->groupsFromDB = [];
```
Comment (ChristophWurst) [inline apps/user_ldap/lib/Jobs/UpdateGroups.php:128]: you could use array_map :)

**Review by ChristophWurst — APPROVED** [review]
Code looks good

File: apps/user_ldap/lib/Jobs/UpdateGroups.php (hunk 1 above)
Comment (blizzz) [reply apps/user_ldap/lib/Jobs/UpdateGroups.php:128]: yes. doesn't make much difference, but perhaps lookin a bit fancier than good ol' foreach.  iirc it is also more memory hungry than foreach.

File: apps/user_ldap/lib/Jobs/UpdateGroups.php [hunk 2]
```diff
@@ -79,28 +114,35 @@ public function updateGroups() {
...
+		foreach ($result as $dataset) {
+			$this->groupsFromDB[$dataset['owncloudname']] = $dataset;
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Jobs/UpdateGroups.php:130]:
```suggestion
			$this->groupsFromDB[$dataset['owncloudname']][]= $dataset;
```
or some other way of merging results? 
Or do you really only care about the last owncloudname user being added? In that case add a group by?

Comment (nickvergessen) [reply apps/user_ldap/lib/Jobs/UpdateGroups.php:130]: Or is owncloudname a unique key?

Comment (blizzz) [reply apps/user_ldap/lib/Jobs/UpdateGroups.php:130]: > Or is owncloudname a unique key?

yes, maps to the groupId

File: apps/user_ldap/lib/Jobs/UpdateGroups.php (hunk 1 above)
Comment (ChristophWurst) [reply apps/user_ldap/lib/Jobs/UpdateGroups.php:128]: who cares about memory footprint if you can have fancy :wink: 

for me it's mostly about readability. If I see an `array_map` I immediately can guess what's happening.

Comment (blizzz) [reply apps/user_ldap/lib/Jobs/UpdateGroups.php:128]: Actually can be a concern, depending how users and groups you throw on it. In this case the readability wouldn't really change as it is a one-line-transformation of the result array, either way.

**Pull request merged**